### PR TITLE
ENG-12120: Add custom EC2 health check script

### PR DESCRIPTION
**Pull request by ccampo133**
Just finished testing this live:

```
$ journalctl -fu healthcheck
...
Sep 15 21:25:02 ip-172-30-0-238.us-east-2.compute.internal systemd[1]: Started Sidecar recycle EC2 health check.
Sep 15 21:25:02 ip-172-30-0-238.us-east-2.compute.internal healthcheck.sh[17181]: Sidecar instance has been marked for recycling - setting EC2 health to 'Unhealthy'
Sep 15 21:25:32 ip-172-30-0-238.us-east-2.compute.internal systemd[1]: Started Sidecar recycle EC2 health check.
Connection to 18.221.73.212 closed by remote host.
Connection to 18.221.73.212 closed.
```

The ASG killed the instance and a new one was created :)

### CODE REVIEW

File: files/cloud-init-pre.sh.tmpl [hunk 1]
```diff
@@ -18,6 +18,47 @@ if [ -z $SIDECAR_VERSION ]; then
...
+[Service]
+ExecStart=/home/ec2-user/cyral/healthcheck.sh
+EOF
+
+  cat << EOF > /etc/systemd/system/healthcheck.timer
+[Unit]
+Description=Sidecar recycle EC2 health check every 30 sec
+[Timer]
+OnBootSec=1min
+OnUnitActiveSec=30
```
Comment (ccampo133) [inline files/cloud-init-pre.sh.tmpl:51]: Should this be a parameter? Or perhaps more frequent than 30 seconds?

Comment (ptcar2009) [reply files/cloud-init-pre.sh.tmpl:51]: I think having this parameterized is good, if possible

File: files/cloud-init-pre.sh.tmpl [hunk 2]
```diff
@@ -18,6 +18,47 @@ if [ -z $SIDECAR_VERSION ]; then
     echo "Sidecar version is undefined."
     exit 1
   fi
+
+  # Create custom health check scripts
+  cat << EOF > /home/ec2-user/cyral/healthcheck.sh
+#!/usr/bin/env bash
+token=\$(${curl} -sf -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600")
+instance_id=\$(${curl} -sf -H "X-aws-ec2-metadata-token: \$token" http://169.254.169.254/latest/meta-data/instance-id || echo "\$HOSTNAME")
```
Comment (wcmjunior) [inline files/cloud-init-pre.sh.tmpl:26]: We have the instance ID stored in our `.env` file. Loading it instead of be an option?

Comment (ptcar2009) [inline files/cloud-init-pre.sh.tmpl:26]: Do we not already have this instance ID? How do we discover this in the sidecar services?

File: files/cloud-init-pre.sh.tmpl [hunk 3]
```diff
@@ -18,6 +18,47 @@ if [ -z $SIDECAR_VERSION ]; then
...
+health=\$(${curl} -sf -X GET "http://localhost:9000/health")
```
Comment (ptcar2009) [inline files/cloud-init-pre.sh.tmpl:27]: Does this work if we have a docker network (not shared with the machine)? If not, how would we solve this?

File: files/cloud-init-pre.sh.tmpl [hunk 4]
```diff
@@ -18,6 +18,47 @@ if [ -z $SIDECAR_VERSION ]; then
...
+if [ \$? -ne 0 ]; then
+  echo "Error getting sidecar health"
+  exit 1
+fi
+recycle_status=\$(echo \$health | jq -r .components.recycle.status)
+if [ \$recycle_status == "degraded" ]; then
+  echo "Sidecar instance has been marked for recycling - setting EC2 health to 'Unhealthy'"
+  aws autoscaling set-instance-health --region ${aws_region} --instance-id \$instance_id --health-status Unhealthy
```
Comment (ptcar2009) [inline files/cloud-init-pre.sh.tmpl:35]: maybe an error check here?

File: files/cloud-init-pre.sh.tmpl (hunk 2 above)
Comment (ccampo133) [reply files/cloud-init-pre.sh.tmpl:26]: It gets set a few lines below. It requires a call to the `fetch_hostname` function. I can just have that function export it as a global env var and use that.

File: files/cloud-init-pre.sh.tmpl (hunk 3 above)
Comment (ccampo133) [reply files/cloud-init-pre.sh.tmpl:27]: It probably does not work. I didn't think of that. We only use the network for e2e tests so far as I understand.

File: files/cloud-init-pre.sh.tmpl (hunk 4 above)
Comment (ccampo133) [reply files/cloud-init-pre.sh.tmpl:35]: Don't need it - if it fails, there's nothing we can do and the script will exit.

File: files/cloud-init-pre.sh.tmpl (hunk 2 above)
Comment (ccampo133) [reply files/cloud-init-pre.sh.tmpl:26]: Let's move the discussion to https://github.com/cyralinc/terraform-aws-sidecar-ec2/pull/83/files#r1327825219

File: files/cloud-init-pre.sh.tmpl (hunk 3 above)
Comment (wcmjunior) [reply files/cloud-init-pre.sh.tmpl:27]: We don't have docker networks and we won't have as we move to single container, so I don't see this as a problem.

File: files/cloud-init-pre.sh.tmpl (hunk 1 above)
Comment (wcmjunior) [reply files/cloud-init-pre.sh.tmpl:51]: I hate the lack of upvote on GitHub, but here it goes: +1

**Review by wcmjunior — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**